### PR TITLE
Allow for the 'file' parameter in SinglefileData.set_file to be a pathlib.Path.

### LINKED ISSUE #4570: pathlib support for `file` constructor arguments
### Is your feature request related to a problem? Please describe

```python
In [1]: from aiida.orm import SinglefileData

In [2]: !pwd
/Users/leopold

In [3]: s=SinglefileData(file='/Users/leopold/nanopub.zip')

In [4]: from pathlib import Path

In [5]: s=SinglefileData(file=Path('/Users/leopold/nanopub.zip'))
-------------------------------------------------------------------------
AttributeError                          Traceback (most recent call last)
<ipython-input-5-e5ace348d29f> in <module>
----> 1 s=SinglefileData(file=Path('/Users/leopold/nanopub.zip'))

~/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/orm/nodes/data/singlefile.py in __init__(self, file, filename, **kwargs)
     45             if filename is None:
     46                 # don't assume that set_file has a 'filename' argument (remove guard in 2.0.0)
---> 47                 self.set_file(file)
     48             else:
     49                 self.set_file(file, filename=filename)

~/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/orm/nodes/data/singlefile.py in set_file(self, file, filename)
    130
    131         if is_filelike:
--> 132             self.put_object_from_filelike(file, key, mode='wb')
    133         else:
    134             self.put_object_from_file(file, key)

~/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/orm/nodes/node.py in put_object_from_filelike(self, handle, path, mode, encoding, force, key)
    637             raise TypeError("put_object_from_filelike() missing 1 required positional argument: 'path'")
    638
--> 639         self._repository.put_object_from_filelike(handle, path, mode, encoding, force)
    640
    641     def delete_object(self, path=None, force=False, key=None):

~/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/orm/utils/_repository.py in put_object_from_filelike(self, handle, key, mode, encoding, force)
    229             folder = folder.get_subfolder(basepath, create=True)
    230
--> 231         folder.create_file_from_filelike(handle, key, mode=mode, encoding=encoding)
    232
    233     def delete_object(self, key, force=False):

~/Personal/Postdoc-MARVEL/repos/aiida/aiida_rmq/aiida/common/folders.py in create_file_from_filelike(self, filelike, filename, mode, encoding)
    226
    227         with open(filepath, mode=mode, encoding=encoding) as handle:
--> 228             shutil.copyfileobj(filelike, handle)
    229
    230         os.chmod(filepath, self.mode_file)

~/Applications/miniconda3/envs/aiida_rmq_py3/lib/python3.7/shutil.py in copyfileobj(fsrc, fdst, length)
     77     """copy data from file-like object fsrc to file-like object fdst"""
     78     while 1:
---> 79         buf = fsrc.read(length)
     80         if not buf:
     81             break

AttributeError: 'PosixPath' object has no attribute 'read'
```

### Describe the solution you'd like

We should start supporting `pathlib.Path` objects, wherever we accept paths in string form.

### PR DESCRIPTION
**Pull request by greschd**
Fixes #4570.

The ``pathlib`` module was introduced in python 3.4 to make path wrangling easier.

The only reason a ``pathlib.Path`` can not currently be used in ``SinglefileData`` is the ``isinstance`` check for ``str`` to determine if the input is a ``filelike`` or a path.

### CODE REVIEW

**Comment by greschd** [comment]
Ah, the ``os`` module was changed to accept ``pathlib`` objects only in 3.6 -- so this will have to wait.

**Comment by greschd** [comment]
A clean rebase and green tests after almost 1 year - that's gotta be some kind of miracle.

File: tests/orm/data/test_singlefile.py [hunk 1]
```diff
@@ -105,6 +106,45 @@ def test_construct_from_string(self):
...
+        self.assertTrue(isinstance(node_loaded, SinglefileData))
+
+        with node.open() as handle:
+            content_loaded = handle.read()
+
+        self.assertEqual(content_loaded, content_original)
+        self.assertEqual(node_loaded.list_object_names(), [basename])
+
+        with node_loaded.open() as handle:
+            self.assertEqual(handle.read(), content_original)
```
Comment (ltalirz) [inline tests/orm/data/test_singlefile.py:146]: Is this copied from another test?

Do we need to test this again?

**Review by ltalirz — APPROVED** [review]
thanks @greschd !

the test is quite extensive and seems to test a couple of things that should not be affected by this change - I leave it up to you whether you prefer to keep it like this (otherwise, please merge)

**Comment by ltalirz** [comment]
> A clean rebase and green tests after almost 1 year - that's gotta be some kind of miracle.

God wants us to introduce pathlib, it's undeniable

**Review by ltalirz — DISMISSED** [review]

File: tests/orm/data/test_singlefile.py (hunk 1 above)
Comment (greschd) [reply tests/orm/data/test_singlefile.py]: Yeah that was copied from the surrounding tests - there was actually quite a bit of duplication in these tests. Couldn't help myself and refactored that into two fixtures.

**Review by ltalirz — APPROVED** [review]
thanks a lot @greschd , much appreciated!

**Comment by greschd** [comment]
Thanks for the review @ltalirz - you'll have to do the merging, since I can't.

**Pull request merged**